src/providers/http.rs

```diff
@@ -0,0 +1,261 @@
+                reqwest::header::HeaderValue::from_str(&format!("Basic {encoded}")).unwrap(),
+            );
+        }
+
+        let base_url = reqwest::Url::from_str(&format!(
+            "{}://{}/{}",
+            if is_secure { "https" } else { "http" },
+            endpoint,
+            API_PATH
+        ))
+        .unwrap();
+
+        Self {
+            inner: reqwest::Client::new(),
+            headers,
+            base_url,
+        }
+    }
+}
+
+const ETHEREUM_HEIGHT_PATH: &str = "eth/height";
+const ETHEREUM_HEADERS_PATH: &str = "eth/headers";
+
+#[async_trait]
+impl EthereumProvider for HttpProviderImpl {
+    async fn get_height(&self) -> Result<u64> {
+        let url = self.url(ETHEREUM_HEIGHT_PATH)?;
+
+        let height = self
+            .inner
+            .get(url)
+            .headers(self.headers.clone())
+            .send()
+            .await?
+            .error_for_status()?
+            .text()
+            .await?
+            .parse()
+            .map_err(|_| Error::Custom("Failed to parse height".to_string()))?;
+
+        Ok(height)
+    }
+
+    async fn get_headers(&self, opts: QueryOptions) -> StreamResponse<BlockHeader> {
+        let url = self.url(ETHEREUM_HEADERS_PATH)?;
+        self.request(url, opts).await
+    }
+}
+
+const UNISWAP_V2_PAIRS_PATH: &str = "uniswap/v2/pairs";
+const UNISWAP_V2_PAIR_PRICES_PATH: &str = "prices";
+const UNISWAP_V2_PAIR_RESERVES_PATH: &str = "raw/reserves";
+const UNISWAP_V2_PAIR_TRADES_PATH: &str = "trades";
+
+#[async_trait]
+impl UniswapV2Provider for HttpProviderImpl {
+    async fn get_pairs(&self, opts: QueryOptions) -> StreamResponse<uniswap_v2::Pair> {
+        let url = self.url(UNISWAP_V2_PAIRS_PATH)?;
+        self.request(url, opts).await
+    }
+
+    async fn get_pair(&self, pair_address: H160) -> Result<uniswap_v2::Pair> {
+        let url = self.url(&format!("{UNISWAP_V2_PAIRS_PATH}/{pair_address:#x}"))?;
+        let opts = QueryOptions::default().start(Bound::Exact(0));
+        let stream = self.request::<uniswap_v2::Pair>(url, opts).await?;
+        futures::pin_mut!(stream);
+
+        let Some(pair) = stream.next().await else {
+            return Err(Error::Custom(format!("Pair {pair_address} not found")));
+        };
+
+        match pair {
+            Ok(pair) => Ok(pair),
+            // fixme - this is not a good way to handle this
+            Err(err) => Err(Error::Custom(format!("Failed to get pair: {err}"))),
+        }
+    }
+
+    async fn get_prices<I>(&self, pairs: I, opts: QueryOptions) -> StreamResponse<uniswap_v2::Price>
+    where
+        I: IntoIterator<Item = H160> + Send,
+    {
+        let pairs: Vec<H160> = pairs.into_iter().collect();
+        if pairs.is_empty() {
+            return Err(Error::Custom("No pairs provided".to_string()));
+        }
+
+        let url = self.url(&format!(
+            "{}/{:#x}/{}",
+            UNISWAP_V2_PAIRS_PATH, pairs[0], UNISWAP_V2_PAIR_PRICES_PATH
+        ))?;
```

Comment:
The `if` condition could be replaced with `match` or `let [pair] = pairs else { ... };`
I think we should error in case the user provides too many pairs to prevent confusion. (same for other methods)
```suggestion
        let pair = match &pairs {
            [pair] => pair,
            [] => return Err(Error::Custom("No pairs provided".to_string())),
            _ => return Err(Error::Custom("Http does not yet allow for multi-pair-queries".to_string())),
        };

        let url = self.url(&format!("{UNISWAP_V2_PAIRS_PATH}/{pair:#x}/{UNISWAP_V2_PAIR_PRICES_PATH}"))?;
```